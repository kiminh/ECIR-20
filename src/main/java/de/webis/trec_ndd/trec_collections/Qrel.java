package de.webis.trec_ndd.trec_collections;

import java.io.Serializable;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

@Data
@NoArgsConstructor
@Accessors(chain = true)
@SuppressWarnings("serial")
public class Qrel implements Serializable{
    private int score;
    private int topicNumber;
    private String documentID;
    
	public Qrel(String line) {
        String[] args = line.split("[\\s\\t]+");
        
        if (args.length == 5) {
        	//prels, like: 1 clueweb09-en0003-55-31884 0 0 1
            topicNumber = Integer.parseInt(args[0]);
            documentID = args[1];
            //FIXME: what is the score for prels?
            score = Integer.MIN_VALUE;
        }
        else {
            topicNumber = Integer.parseInt(args[0]);
            //skipping 1 because it is irrelevant -> has always 0 as value wehen printed
            documentID = args[2];
            score = Integer.parseInt(args[3]);	
        }
    }
	

	
	public String toString() {
		return new StringBuilder().append(topicNumber).append(" 0 ").append(documentID).append(' ').append(score).toString();
	}
	
	@Override
	public int hashCode() {
		return toString().hashCode();
	}
	
	@Override
	public boolean equals(Object o) {
		if(o instanceof Qrel) {
			return this.documentID.equals(((Qrel) o).documentID) && this.score==((Qrel)o).score && this.topicNumber==((Qrel)o).topicNumber;
		}
		else return false;
	}
}

