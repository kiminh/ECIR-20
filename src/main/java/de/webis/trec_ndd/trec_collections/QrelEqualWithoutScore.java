package de.webis.trec_ndd.trec_collections;

import java.io.Serializable;

import lombok.experimental.Accessors;

@Accessors(chain = true)
@SuppressWarnings("serial")
public class QrelEqualWithoutScore extends Qrel implements Serializable{

	public QrelEqualWithoutScore() {
		
	}
	
	public QrelEqualWithoutScore(String line) {
		super(line);
	}
	
	@Override
	public boolean equals(Object o) {
		if(o instanceof Qrel || o instanceof QrelEqualWithoutScore) {
			return this.getDocumentID().equals(((Qrel) o).getDocumentID()) && this.getTopicNumber()==((Qrel)o).getTopicNumber();
		}
		else return false;
	}
	
	@Override
	public int hashCode() {
		return new StringBuilder().append(getTopicNumber()).append(getDocumentID()).toString().hashCode();
	}
	
	@Override
	public String toString() {
		return super.toString();
	}
	
	public QrelEqualWithoutScore copy() {
		QrelEqualWithoutScore ret = new QrelEqualWithoutScore();
		ret.setDocumentID(this.getDocumentID());
		ret.setTopicNumber(this.getTopicNumber());
		ret.setScore(this.getScore());
		
		return ret;
	}
	
}