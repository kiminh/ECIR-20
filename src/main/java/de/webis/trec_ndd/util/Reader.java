package de.webis.trec_ndd.util;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;

import lombok.SneakyThrows;

public class Reader<T> implements Iterator<T>, Iterable<T>{
	
	private Class<T> readerType;
	private BufferedReader[] readerList;
	private BufferedReader currentReader;
	private int currentReaderIndex=0;
	private String lastLine=null;
	private File fileToBeRead;
	
	private Reader(Reader<T> r) throws IOException{
		init(r.fileToBeRead, r.readerType);
	}

	
	public Reader(File file, Class<T> cls) throws IOException{
		init(file,cls);
	}
	
	public Reader(String sfile, Class<T> cls) throws IOException{
		init(new File(sfile),cls);
	}
	
	public void init(File file, Class<T> cls) throws IOException{
		this.fileToBeRead=file;
		readerType=cls;
		if(!file.exists()) throw new FileNotFoundException();
		if(file.isDirectory()) {
			File[] fileList;
			if((fileList=file.listFiles())==null || fileList.length==0) throw new IOException("no groups found");
			this.readerList=new BufferedReader [fileList.length];
			for(int i=0; i<fileList.length; i++) {
				this.readerList[i]=new BufferedReader(new FileReader(fileList[i]));
			}
		}else {
			this.readerList=new BufferedReader[1];
			this.readerList[0]=new BufferedReader(new FileReader(file));
		}
		this.currentReader=this.readerList[0];
		this.lastLine=this.currentReader.readLine();
	}
	
	@Override
	public boolean hasNext() {
		return lastLine!=null;
	}

	@Override
	@SneakyThrows
	public T next() {
		T res=this.readerType.getDeclaredConstructor(String.class).newInstance(this.lastLine);
		this.lastLine=currentReader.readLine();
		while(this.lastLine==null && this.currentReaderIndex+1<this.readerList.length) {
			this.currentReader=this.readerList[++this.currentReaderIndex];
			this.lastLine=currentReader.readLine();
		}
		return res;
	}

	@Override
	public Iterator<T> iterator() {
		// TODO Auto-generated method stub
		return this;
	}
	
	@SneakyThrows
	public void close() {
		for(int i=0; i<this.readerList.length; i++) {
			this.readerList[i].close();
		}
	}
	
	@SneakyThrows
	public ArrayList<T> getArrayList(){
		ArrayList <T> res = new ArrayList<T>();
		Reader<T> reader = new Reader<T>(this);
		for(T element: reader) {
			res.add(element);
		}
		reader.close();
		return res;
	}
	
	@SneakyThrows
	public Collection<T> getCollection(){
		return (Collection <T>) getArrayList();
	}
}
