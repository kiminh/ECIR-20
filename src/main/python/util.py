def all_files_in_hdfs_dir(hdfs_path):
    import os
    import re
    import subprocess

    cmd = 'hdfs dfs -ls ' + hdfs_path
    cmd_output = subprocess.check_output(cmd, shell=True)
    ret = []
    for line in cmd_output.decode('utf-8').strip().split('\n'):
        fields = re.split('\s+', line)
        if len(fields) == 8:
            ret += [fields[-1]]

    return ret

