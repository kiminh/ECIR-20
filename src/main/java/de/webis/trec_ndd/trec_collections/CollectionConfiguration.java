package de.webis.trec_ndd.trec_collections;

import java.io.InputStream;
import java.io.Serializable;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import de.webis.trec_ndd.trec_collections.SharedTask.TrecSharedTask;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.SneakyThrows;

public interface CollectionConfiguration extends Serializable {

	public String getPathToCollection();
	public String getCollectionType();
	public String getDocumentGenerator();
	public List<SharedTask> getSharedTasks();

	@SneakyThrows
	public default Set<String> judgedDocumentIds() {
		Set<String> ret = new HashSet<>();
		
		for(String qrelResource : getQrelResourcesOfAllSharedTasks()) {
			InputStream resource = AnseriniCollectionReader.class.getResourceAsStream(qrelResource);
			ret.addAll(new QrelReader().readJudgedDocumentIdsFromQrels(resource));
		}
		
		return ret;
	}
	
	public default HashSet<String> documentIdsInRunFiles() {
		HashSet<String> ret = new HashSet<>();
		
		for(SharedTask sharedTask: getSharedTasks()) {
			ret.addAll(sharedTask.documentIdsInRunFiles());
		}
		
		return ret;
	}
	
	public default List<String> getQrelResourcesOfAllSharedTasks() {
		return getSharedTasks().stream()
				.map(SharedTask::getQrelResource)
				.collect(Collectors.toList());
	}
	
	@Getter
	@AllArgsConstructor
	public static enum TrecCollections implements CollectionConfiguration {
		CLUEWEB09(
			// compressed 4TB
			"/mnt/ceph/storage/corpora/corpora-thirdparty/corpus-clueweb09/",
			"ClueWeb09Collection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
				TrecSharedTask.WEB_2009,
				TrecSharedTask.WEB_2010,
				TrecSharedTask.WEB_2011,
				TrecSharedTask.WEB_2012,
				TrecSharedTask.SESSION_2010,
				TrecSharedTask.SESSION_2011,
				TrecSharedTask.SESSION_2012
			))
		),
		CLUEWEB12(
			"/mnt/ceph/storage/data-in-progress/kibi9872/clueweb12/",
			"ClueWeb12Collection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.WEB_2013,
					TrecSharedTask.WEB_2014,
					TrecSharedTask.SESSION_2013,
					TrecSharedTask.SESSION_2014
			))
		),
		GOV1(
			// compressed 4.6 GB
			"/mnt/nfs/webis20/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV/",
			"TrecwebCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Collections.emptyList())
		),
		GOV2(
			// compressed 81GB
			"/mnt/nfs/webis20/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV2/gov2-corpus/",
			"TrecwebCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.TERABYTE_2004,
					TrecSharedTask.TERABYTE_2005_ADHOC,
					TrecSharedTask.TERABYTE_2006
			))
		),
		GOV2_MQ(			// compressed 81GB
				"/mnt/ceph/storage/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV2/gov2-corpus/",
				"TrecwebCollection",
				"JsoupGenerator",
				Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.MILLION_QUERY_2007,
					TrecSharedTask.MILLION_QUERY_2008
				))
		),
		ROBUST04(
			"/mnt/ceph/storage/data-in-progress/kibi9872/robust-04/",
			"TrecCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.ROBUST_04))
		),
		CORE2018(
			"/mnt/nfs/webis20/data-in-progress/trec-2018-webis/WashingtonPost/WashingtonPost.v2/data/",
			"WashingtonPostCollection",
			"WapoGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.CORE_2018))
		),
		CORE2017(
			"/mnt/nfs/webis20/data-in-progress/trec-2017-webis/data/nyt/",
			"NewYorkTimesCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.CORE_2017))
		);
		

		private final String pathToCollection,
							collectionType,
							documentGenerator;

		private final List<SharedTask> sharedTasks;
	}
	
	@Getter
	@AllArgsConstructor
	public static enum OtherCollections implements CollectionConfiguration {
		COMMON_CRAWL_2015_02("CommonCrawlCollection", "JsoupGenerator");
		
		private final String collectionType,
			documentGenerator;

		@Override
		public List<SharedTask> getSharedTasks() {
			return Collections.emptyList();
		}

		@Override
		public String getPathToCollection() {
			return null;
		}
	}
}
