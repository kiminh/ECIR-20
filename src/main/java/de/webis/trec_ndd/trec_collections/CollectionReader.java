package de.webis.trec_ndd.trec_collections;

import java.util.List;

public interface CollectionReader {
	public List<CollectionDocument> extractJudgedRawDocumentsFromCollection();
	
	public List<CollectionDocument> extractRunFileDocumentsFromsCollection();

	public List<CollectionDocument> extractRawDocumentsFromCollection();
}
