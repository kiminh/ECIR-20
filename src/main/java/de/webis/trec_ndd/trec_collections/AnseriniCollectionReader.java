package de.webis.trec_ndd.trec_collections;

import java.io.Serializable;
import java.lang.reflect.InvocationTargetException;
import java.net.URL;
import java.nio.file.Paths;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

import org.apache.lucene.document.Document;

import com.google.common.collect.Iterators;

import de.webis.trec_ndd.spark.DocumentHash;
import de.webis.trec_ndd.trec_collections.CollectionConfiguration.OtherCollections;
import io.anserini.collection.ClueWeb09Collection;
import io.anserini.collection.ClueWeb12Collection;
import io.anserini.collection.CommonCrawlCollection;
import io.anserini.collection.DocumentCollection;
import io.anserini.collection.Segment;
import io.anserini.collection.SegmentProvider;
import io.anserini.collection.SourceDocument;
import io.anserini.index.IndexCollection;
import io.anserini.index.IndexCollection.Args;
import io.anserini.index.IndexCollection.Counters;
import io.anserini.index.generator.LuceneDocumentGenerator;
import lombok.Data;
import lombok.SneakyThrows;

@Data
@SuppressWarnings("serial")
public class AnseriniCollectionReader<T extends SourceDocument> implements CollectionReader, Serializable {
	private final CollectionConfiguration config;
	
	@Override
	public List<CollectionDocument> extractJudgedRawDocumentsFromCollection() {
		List<CollectionDocument> ret = new LinkedList<>();

		for(String segmentPath : segmentPaths()) {
			judgedCollectionDocumentsInPath(segmentPath).forEachRemaining(ret::add);
		}

		return ret;
	}
	
	@Override
	public List<CollectionDocument> extractRunFileDocumentsFromsCollection() {
		List<CollectionDocument> ret = new LinkedList<>();

		for(String segmentPath : segmentPaths()) {
			documentsFromRunFilesInPath(segmentPath).forEachRemaining(ret::add);
		}

		return ret;
	}

	@Override
	public List<CollectionDocument> extractRawDocumentsFromCollection() {
		List<CollectionDocument> ret = new LinkedList<>();

		for(String segmentPath : segmentPaths()) {
			collectionDocumentsInPath(segmentPath).forEachRemaining(ret::add);
		}

		return ret;
	}

	@SneakyThrows
	public List<String> segmentPaths() {
		return documentCollection().getFileSegmentPaths()
				.stream()
				.map(p -> p.toAbsolutePath().toString())
				.collect(Collectors.toList());
	}
	
	public Iterator<DocumentHash> documentHashIterator(String segmentPath) {
		Iterator<CollectionDocument> iter = collectionDocumentsInPath(segmentPath);
		
		return Iterators.transform(iter, d -> new DocumentHash(d));
	}
	
	public Iterator<CollectionDocument> judgedCollectionDocumentsInPath(String segmentPath) {
		Set<String> judgedDocumentIds = config.judgedDocumentIds();
		
		return Iterators.filter(iteratorForCollectionDocumentsInPath(segmentPath, doc -> judgedDocumentIds.contains(documentId(doc))), i -> i != null);
	}
	
	public Iterator<CollectionDocument> documentsFromRunFilesInPath(String segmentPath) {
		Set<String> documentIdsInRunFiles = config.documentIdsInRunFiles();
		
		return documentsFromRunFilesInPath(segmentPath, documentIdsInRunFiles);
	}
	
	public Iterator<CollectionDocument> documentsFromRunFilesInPath(String segmentPath, Set<String> documentIdsInRunFiles) {
		return Iterators.filter(iteratorForCollectionDocumentsInPath(segmentPath, doc -> documentIdsInRunFiles.contains(documentId(doc))), i -> i != null);
	}
	
	public Iterator<CollectionDocument> collectionDocumentsInPath(String segmentPath) {
		return Iterators.filter(iteratorForCollectionDocumentsInPath(segmentPath, null), i -> i != null);
	}
	
	@SneakyThrows
	private Iterator<CollectionDocument> iteratorForCollectionDocumentsInPath(String segmentPath, Function<Document, Boolean> keepDocument) {
		SegmentProvider<T> collection = documentCollection();
		
		return new AnseriniCollectionSegmentIterator<>(
				collection, 
				documentGenerator(),
				collection.createFileSegment(Paths.get(segmentPath)),
				segmentPath,
				keepDocument
		);
	}

	@Data
	private static class AnseriniCollectionSegmentIterator<T extends SourceDocument>
			implements Iterator<CollectionDocument> {
		private final SegmentProvider<T> collection;
		private final LuceneDocumentGenerator<T> generator;
		private final Segment<T> segment;
		private final String segmentPath;
		private final Function<Document, Boolean> keepDocument;

		@Override
		public boolean hasNext() {
			return segment.hasNext();
		}

		@Override
		public CollectionDocument next() {
			T document = segment.next();
			if (document == null || !document.indexable()) {
				return null;
			}
			
			Document doc = generator.createDocument(document);
			if (doc == null || (keepDocument != null && Boolean.FALSE.equals(keepDocument.apply(doc)))) {
				return null;
			} else {
				CollectionDocument ret = CollectionDocument.fromLuceneDocument(doc);
				ret.setUrl(extractUrlIfPossible(document));
				return ret;
			}
		}
	}
	
	private static URL extractUrlIfPossible(SourceDocument document) {
		try {
			return extractUrlOrFail(document);
		} catch (Exception e) {
			return null;
		}
	}

	private static URL extractUrlOrFail(SourceDocument document) throws Exception {
		if (document instanceof ClueWeb09Collection.Document) {
			return new URL(((ClueWeb09Collection.Document) document).getURL());
		} else if (document instanceof ClueWeb09Collection.Document) {
			return new URL(((ClueWeb12Collection.Document) document).getURL());
		}
		
		return null;
	}
	
	@SneakyThrows
	@SuppressWarnings({ "rawtypes", "unchecked" })
	private SegmentProvider<T> documentCollection()
			throws InstantiationException, IllegalAccessException, ClassNotFoundException {
		SegmentProvider<T> ret = (SegmentProvider) Class.forName("io.anserini.collection." + args().collectionClass)
				.newInstance();
		((DocumentCollection) ret).setCollectionPath(Paths.get(config.getPathToCollection()));

		if(ret instanceof CommonCrawlCollection) {
			((CommonCrawlCollection) ret).setS3Files(((OtherCollections) config).getS3Files());
		}
		
		return ret;
	}

	@SneakyThrows
	@SuppressWarnings({ "rawtypes", "unchecked" })
	private LuceneDocumentGenerator<T> documentGenerator() throws InstantiationException, IllegalAccessException,
			IllegalArgumentException, InvocationTargetException, NoSuchMethodException, SecurityException, Exception {
		Args args = args();
		Class<?> clazz = Class.forName("io.anserini.index.generator." + args.generatorClass);

		return (LuceneDocumentGenerator) clazz.getDeclaredConstructor(Args.class, Counters.class).newInstance(args,
				new IndexCollection(args).new Counters());
	}

	private Args args() {
		Args ret = new Args();
		ret.generatorClass = config.getDocumentGenerator();
		ret.collectionClass = config.getCollectionType();
		ret.index = "-index";
		ret.input = config.getPathToCollection();

		return ret;
	}

	static String documentId(Document doc) {
		return doc.getField(LuceneDocumentGenerator.FIELD_ID).stringValue();
	}
}
