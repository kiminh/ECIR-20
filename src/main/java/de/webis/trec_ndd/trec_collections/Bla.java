package de.webis.trec_ndd.trec_collections;

import java.io.File;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.codehaus.jackson.map.ObjectMapper;

import de.webis.trec_ndd.trec_collections.CollectionConfiguration.TrecCollections;
import lombok.Data;

public class Bla {
	@Data
	public static class BLA {
		private final String topicNumber;
		private final String topicQuery;
		private final Map<String, String> documentToJudgment; 
	}
	
	public static void main(String [] args) throws Exception {
		Map<String, Object> ret = new LinkedHashMap<>();
		
		for(SharedTask task : TrecCollections.CLUEWEB09.getSharedTasks()) {
			ret.put(task.toString(), taskStuff(task));
		}
		
		new ObjectMapper().writeValue(new File("clueweb.json"), ret);
	}
	
	private static Map<String, BLA> taskStuff(SharedTask task) {
		Map<String, BLA> ret = new HashMap<>();
		
		for(Map.Entry<String, Map<String, String>> i : task.documentJudgments().getData().entrySet()) {
			String query = task.getQueryForTopic(i.getKey());
			ret.put(i.getKey(), new BLA(i.getKey(), query, i.getValue()));
		}
		
		return ret;
	}
}
