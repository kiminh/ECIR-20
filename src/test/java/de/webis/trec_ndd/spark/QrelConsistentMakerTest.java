package de.webis.trec_ndd.spark;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

import org.approvaltests.Approvals;
import org.junit.Test;

import de.webis.trec_ndd.trec_collections.Qrel;
import de.webis.trec_ndd.trec_collections.QrelEqualWithoutScore;
import de.webis.trec_ndd.util.Reader;
import lombok.SneakyThrows;

public class QrelConsistentMakerTest {

	
	private DocumentGroup group;
	private HashMap<String,ArrayList<Qrel>> idMapQrel;
	private static String line = "{\"hash\":\"ac2d11c5d7f77cdef38f1566f3df8b05\",\"ids\":[\"clueweb09-ja0004-63-23900\",\"clueweb09-ja0009-69-49575\",\"clueweb09-ja0014-01-20176\"]}";
	
	public QrelConsistentMakerTest(){
		group =  new DocumentGroup(line);
		idMapQrel = new HashMap<String,ArrayList<Qrel>>();
		idMapQrel.put(group.ids.get(0),new ArrayList<Qrel>());
		idMapQrel.get(group.ids.get(0)).add(new Qrel().setDocumentID(group.ids.get(0)).setTopicNumber(1).setScore(4));
		idMapQrel.get(group.ids.get(0)).add(new Qrel().setDocumentID(group.ids.get(0)).setTopicNumber(2).setScore(10));
		idMapQrel.put(group.ids.get(1),new ArrayList<Qrel>());
		idMapQrel.get(group.ids.get(1)).add(new Qrel().setDocumentID(group.ids.get(1)).setTopicNumber(11).setScore(6));
		idMapQrel.get(group.ids.get(1)).add(new Qrel().setDocumentID(group.ids.get(1)).setTopicNumber(1).setScore(8));
		
	}
	
	@Test
	public void testMaxValue() {
		QrelConsistentMakerTest instance = new QrelConsistentMakerTest();
		HashMap<Integer,Integer> topicMapValue = QrelConsistentMaker.Internals.maxValue(instance.group,instance.idMapQrel);
		Approvals.verifyAsJson(topicMapValue);
	}
	
	@Test
	public void testMinValue() {
		QrelConsistentMakerTest instance = new QrelConsistentMakerTest();
		HashMap<Integer,Integer> topicMapValue = QrelConsistentMaker.Internals.minValue(instance.group,instance.idMapQrel);
		Approvals.verifyAsJson(topicMapValue);
	}
	
	@Test
	public void testAvgValue() {
		QrelConsistentMakerTest instance = new QrelConsistentMakerTest();
		HashMap<Integer,Integer> topicMapValue = QrelConsistentMaker.Internals.avgValue(instance.group,instance.idMapQrel);
		Approvals.verifyAsJson(topicMapValue);
	}
	
	@Test
	public void testCreateMapById() {
		ArrayList<Qrel> qrelList = new ArrayList<Qrel>();
		qrelList.add(new Qrel().setDocumentID("doc1").setTopicNumber(1).setScore(10));
		qrelList.add(new Qrel().setDocumentID("doc1").setTopicNumber(2).setScore(2));
		qrelList.add(new Qrel().setDocumentID("doc1").setTopicNumber(3).setScore(3));
		qrelList.add(new Qrel().setDocumentID("doc2").setTopicNumber(1).setScore(6));
		HashMap<String, ArrayList<Qrel>> idMapQrelTest = QrelConsistentMaker.Internals.createMapById(qrelList);
		Approvals.verifyAsJson(idMapQrelTest);
	}
	

	
	//TODO 
	//test all configs
	@Test
	@SneakyThrows
	public void testWriteChangedQrel() {
		HashSet<QrelEqualWithoutScore> judgements = new HashSet<QrelEqualWithoutScore>();
		judgements.addAll(new Reader<QrelEqualWithoutScore>("./src/test/resources/QrelConsistentMaker/qrelFile", QrelEqualWithoutScore.class).getCollection());
		List<DocumentGroup> groups = new Reader<DocumentGroup>("./src/test/resources/QrelConsistentMaker/groupFile", DocumentGroup.class).getArrayList();
		Approvals.verifyAsJson(QrelConsistentMaker.base.getQrels(judgements,groups, null));
	}
	
	
}
