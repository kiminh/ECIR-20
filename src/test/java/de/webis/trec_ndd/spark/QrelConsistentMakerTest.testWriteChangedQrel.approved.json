[
  {
    "score": 100,
    "topicNumber": 2,
    "documentID": "clueweb09-zh0036-08-11760"
  },
  {
    "score": 0,
    "topicNumber": 1,
    "documentID": "clueweb09-zh0009-35-18959"
  },
  {
    "score": 111,
    "topicNumber": 1,
    "documentID": "clueweb09-zh0009-35-18962"
  },
  {
    "score": 10,
    "topicNumber": 1,
    "documentID": "clueweb09-en0069-96-24946"
  },
  {
    "score": 6,
    "topicNumber": 1,
    "documentID": "clueweb09-en0064-73-24946"
  }
]